src/echo/echo.rs

```diff
@@ -36,119 +36,90 @@ static HELP: &str = r#"
  \\xHH    byte with hexadecimal value HH (1 to 2 digits)
 "#;
 
-enum Base {
-    B8 = 8,
-    B16 = 16,
-}
-
-struct Opts {
-    newline: bool,
-    escape: bool,
+fn parse_code(input: &mut Peekable<Chars>, base: u32, max_digits: u32, bits_per_digit: u32) -> Option<char> {
+    let mut ret = 0x80000000;
+    for _ in 0..max_digits {
+        match input.peek().and_then(|c| c.to_digit(base)) {
+            Some(n) => ret = (ret << bits_per_digit) | n,
+            None => break,
+        }
+        input.next();
+    }
+    std::char::from_u32(ret)
 }
 
-fn convert_str(string: &[u8], index: usize, base: Base) -> (char, usize) {
-    let (max_digits, is_legal_digit): (usize, fn(u8) -> bool) = match base {
-        Base::B8 => (3, |c| (c as char).is_digit(8)),
-        Base::B16 => (2, |c| (c as char).is_digit(16)),
-    };
-
-    let mut bytes = vec![];
-    for offset in 0..max_digits {
-        if string.len() <= index + offset as usize {
-            break;
-        }
-        let c = string[index + offset as usize];
-        if is_legal_digit(c) {
-            bytes.push(c as u8);
-        } else {
-            break;
+pub fn unescape(input: &str) -> String {
```

Comment:
To be honest, I'm not entirely sold that we should create a new `String` as the result.  I still think it might be better to lock `stdout` and then just write the output directly to it so that `echo` won't fail on giant inputs in constrained environments.

Reply:
The advantage would be that this is more easily unit testable and keeps the function pure. I had assumed allocation efficiency was a non-goal as the previous implementation didn't seem to make much of an effort to avoid allocations.

Reply:
The only thing that allocated a significant amount was `matches.free.join()`.  Preferably even that wouldn't be present, as we'd just print a space character when we loop over the arguments.

One solution to make this unit testable would be to make the function generic over `Write`.  You should be able to pass in a buffer and test the contents of the buffer if you do so.